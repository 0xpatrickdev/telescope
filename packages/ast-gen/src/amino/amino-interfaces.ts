import * as t from '@babel/types';
import { snake } from 'case';
import { Enum, Field, Interface, MessageSchema } from '../types';
import { toAmino } from './to-amino';
import { fromAmino } from './from-amino';
import { getFieldDimensionality, typeUrlToAmino, arrayTypeNDimensions } from '../utils';

export const makeAminoConverterItem = (schema: MessageSchema, enums: Enum[], interfaces: Interface[]) => {
  return t.objectProperty(
    t.stringLiteral(schema.typeUrl),
    t.objectExpression(
      [
        t.objectProperty(t.identifier('aminoType'), t.stringLiteral(typeUrlToAmino(schema.typeUrl, schema.name))),
        t.objectProperty(t.identifier('toAmino'), toAmino(schema, enums, interfaces)),
        t.objectProperty(t.identifier('fromAmino'), fromAmino(schema, enums, interfaces))
      ]
    )
  );
};

export const aminoConverter = (schemas: MessageSchema[], enums: Enum[], interfaces: Interface[]) => {
  return t.exportNamedDeclaration(t.variableDeclaration('const', [
    t.variableDeclarator(t.identifier('AminoConverter'),
      t.objectExpression(
        schemas.map(schema => makeAminoConverterItem(schema, enums, interfaces))
      ))
  ]));
};

export const renderAminoField = (field: Field, enums: Enum[], interfaces: any[] = [], imports: any[] = []) => {

  const { typeName, dimensions, isArray } = getFieldDimensionality(field);

  // special cases
  switch (field.type) {
    case 'Height':
      // Height becomes AminoHeight...
      imports.push({
        identifier: 'AminoHeight',
      });
      return t.tsPropertySignature(
        t.identifier(snake(field.name)),
        t.tsTypeAnnotation(
          t.tsTypeReference(t.identifier('AminoHeight'))
        )
      );
    default:
  }


  // enums
  const val = enums.find(e => e.name === typeName);
  if (val) {
    if (isArray) {
      return t.tsPropertySignature(
        t.identifier(snake(field.name)),
        t.tsTypeAnnotation(arrayTypeNDimensions(t.tSNumberKeyword(), dimensions))
      );
    }

    return t.tsPropertySignature(
      t.identifier(snake(field.name)),
      t.tsTypeAnnotation(t.tSNumberKeyword())
    );
  }

  // definitions/interfaces
  const ival: Interface = interfaces.find(e => e.name === typeName);
  if (ival) {
    const properties = ival.fields.map(field =>
      renderAminoField(field, enums, interfaces, imports)
    );

    if (isArray) {
      return t.tsPropertySignature(
        t.identifier(snake(field.name)),
        t.tsTypeAnnotation(
          arrayTypeNDimensions(t.tsTypeLiteral(
            properties
          ), dimensions)
        )
      )
    }

    return t.tsPropertySignature(
      t.identifier(snake(field.name)),
      t.tsTypeAnnotation(
        t.tsTypeLiteral(
          properties
        )
      )
    );
  }

  // basic scalar types
  switch (field.type) {
    case 'Long':
      // longs become strings...
      return t.tsPropertySignature(
        t.identifier(snake(field.name)),
        t.tsTypeAnnotation(t.tSStringKeyword())
      );
    default:
      return t.tsPropertySignature(
        t.identifier(snake(field.name)),
        t.tsTypeAnnotation(field.node)
      );
  }
};

export const makeAminoTypeInterface = (schema: MessageSchema, enums: Enum[], interfaces: any[] = [], imports: any[] = []) => {

  const TypeName = schema.name;
  const typeUrl = schema.typeUrl;

  return t.exportNamedDeclaration(
    t.tsInterfaceDeclaration(
      t.identifier('Amino' + TypeName),
      null,
      [t.tsExpressionWithTypeArguments(t.identifier('AminoMsg'))],
      t.tSInterfaceBody([
        t.tSPropertySignature(t.identifier('type'), t.tsTypeAnnotation(
          t.tSLiteralType(t.stringLiteral(typeUrlToAmino(typeUrl, TypeName)))
        )),
        t.tSPropertySignature(t.identifier('value'), t.tsTypeAnnotation(t.tsTypeLiteral(
          schema.fields.map((field) => {
            return renderAminoField(field, enums, interfaces, imports);
          })
        )))
      ])
    )
  )

}