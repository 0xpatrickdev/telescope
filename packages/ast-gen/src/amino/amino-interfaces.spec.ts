import { aminoConverter, makeAminoConverterItem, makeAminoTypeInterface } from './amino-interfaces';
import generate from '@babel/generator';
import { schemata, enums, interfaces } from '../__fixtures__';
const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

it('makeAminoConverterItem', async () => {
    expectCode(makeAminoConverterItem(schemata[0], enums, interfaces));
});


it('makeAminoTypeInterface', async () => {
    expectCode(makeAminoTypeInterface(schemata[0], enums, interfaces));
});

it('aminoConverter', async () => {
    expectCode(aminoConverter(schemata, enums, interfaces));
});