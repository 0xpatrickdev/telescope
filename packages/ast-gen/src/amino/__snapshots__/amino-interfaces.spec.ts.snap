// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aminoConverter 1`] = `
"export const AminoConverter = {
  \\"/cosmos.some.MsgThing\\": {
    aminoType: \\"cosmos-sdk/MstThing\\",
    toAmino: ({
      durationCamelCase,
      coinsCamelCase,
      someDuration,
      someHeight,
      myCoin,
      camelCaseName,
      voteValue,
      previousVotes,
      desc,
      commission,
      str
    }: MstThing): AminoMstThing[\\"value\\"] => {
      return {
        duration_camel_case: durationCamelCase.toString(),
        coins_camel_case: coinsCamelCase,
        some_duration: (someDuration * 1_000_000_000).toString(),
        some_height: someHeight ? {
          revision_height: omitDefault(someHeight.revisionHeight)?.toString(),
          revision_number: omitDefault(someHeight.revisionNumber)?.toString()
        } : {},
        my_coin: {
          denom: myCoin.denom,
          amount: Long.fromNumber(myCoin.amount).toString()
        },
        camel_case_name: camelCaseName,
        vote_value: voteValue,
        previous_votes: previousVotes,
        desc: {
          moniker: desc.moniker,
          identity: desc.identity,
          website: desc.website,
          security_contact: desc.securityContact,
          details: desc.details,
          super_nested: {
            rate: desc.superNested.rate,
            max_rate: desc.superNested.maxRate,
            max_change_rate: desc.superNested.maxChangeRate,
            special_property_here: desc.superNested.specialPropertyHere.map(el0 => ({
              rate: el0.rate,
              max_rate: el0.maxRate,
              max_change_rate: el0.maxChangeRate
            }))
          },
          many_comissions: desc.manyComissions.map(el0 => ({
            rate: el0.rate,
            max_rate: el0.maxRate,
            max_change_rate: el0.maxChangeRate,
            special_property_here: el0.specialPropertyHere.map(el1 => ({
              rate: el1.rate,
              max_rate: el1.maxRate,
              max_change_rate: el1.maxChangeRate
            }))
          }))
        },
        commission: {
          rate: commission.rate,
          max_rate: commission.maxRate,
          max_change_rate: commission.maxChangeRate,
          special_property_here: commission.specialPropertyHere.map(el0 => ({
            rate: el0.rate,
            max_rate: el0.maxRate,
            max_change_rate: el0.maxChangeRate
          }))
        },
        str
      };
    },
    fromAmino: ({
      duration_camel_case,
      coins_camel_case,
      some_duration,
      some_height,
      my_coin,
      camel_case_name,
      vote_value,
      previous_votes,
      desc,
      commission,
      str
    }: AminoMstThing[\\"value\\"]): MstThing => {
      return {
        durationCamelCase: Long.fromString(duration_camel_case),
        coinsCamelCase: Array.from(coins_camel_case),
        someDuration: {
          seconds: Long.fromNumber(Math.floor(parseInt(some_duration) / 1_000_000_000)),
          nanos: parseInt(some_duration) % 1_000_000_000
        },
        someHeight: some_height ? {
          revisionHeight: Long.fromString(some_height.revision_height || \\"0\\", true),
          revisionNumber: Long.fromString(some_height.revision_number || \\"0\\", true)
        } : undefined,
        myCoin: my_coin,
        camelCaseName: camel_case_name,
        voteValue: voteOptionToJSON(vote_value),
        previousVotes: previous_votes.map(el => voteOptionToJSON(el)),
        desc: {
          moniker: desc.moniker,
          identity: desc.identity,
          website: desc.website,
          securityContact: desc.security_contact,
          details: desc.details,
          superNested: {
            rate: desc.super_nested.rate,
            maxRate: desc.super_nested.max_rate,
            maxChangeRate: desc.super_nested.max_change_rate,
            specialPropertyHere: desc.super_nested.special_property_here.map(el2 => ({
              rate: el2.rate,
              maxRate: el2.max_rate,
              maxChangeRate: el2.max_change_rate
            }))
          },
          manyComissions: desc.many_comissions.map(el1 => ({
            rate: el1.rate,
            maxRate: el1.max_rate,
            maxChangeRate: el1.max_change_rate,
            specialPropertyHere: el1.special_property_here.map(el2 => ({
              rate: el2.rate,
              maxRate: el2.max_rate,
              maxChangeRate: el2.max_change_rate
            }))
          }))
        },
        commission: {
          rate: commission.rate,
          maxRate: commission.max_rate,
          maxChangeRate: commission.max_change_rate,
          specialPropertyHere: commission.special_property_here.map(el1 => ({
            rate: el1.rate,
            maxRate: el1.max_rate,
            maxChangeRate: el1.max_change_rate
          }))
        },
        str
      };
    }
  }
};"
`;

exports[`makeAminoConverterItem 1`] = `
"\\"/cosmos.some.MsgThing\\": {
  aminoType: \\"cosmos-sdk/MstThing\\",
  toAmino: ({
    durationCamelCase,
    coinsCamelCase,
    someDuration,
    someHeight,
    myCoin,
    camelCaseName,
    voteValue,
    previousVotes,
    desc,
    commission,
    str
  }: MstThing): AminoMstThing[\\"value\\"] => {
    return {
      duration_camel_case: durationCamelCase.toString(),
      coins_camel_case: coinsCamelCase,
      some_duration: (someDuration * 1_000_000_000).toString(),
      some_height: someHeight ? {
        revision_height: omitDefault(someHeight.revisionHeight)?.toString(),
        revision_number: omitDefault(someHeight.revisionNumber)?.toString()
      } : {},
      my_coin: {
        denom: myCoin.denom,
        amount: Long.fromNumber(myCoin.amount).toString()
      },
      camel_case_name: camelCaseName,
      vote_value: voteValue,
      previous_votes: previousVotes,
      desc: {
        moniker: desc.moniker,
        identity: desc.identity,
        website: desc.website,
        security_contact: desc.securityContact,
        details: desc.details,
        super_nested: {
          rate: desc.superNested.rate,
          max_rate: desc.superNested.maxRate,
          max_change_rate: desc.superNested.maxChangeRate,
          special_property_here: desc.superNested.specialPropertyHere.map(el0 => ({
            rate: el0.rate,
            max_rate: el0.maxRate,
            max_change_rate: el0.maxChangeRate
          }))
        },
        many_comissions: desc.manyComissions.map(el0 => ({
          rate: el0.rate,
          max_rate: el0.maxRate,
          max_change_rate: el0.maxChangeRate,
          special_property_here: el0.specialPropertyHere.map(el1 => ({
            rate: el1.rate,
            max_rate: el1.maxRate,
            max_change_rate: el1.maxChangeRate
          }))
        }))
      },
      commission: {
        rate: commission.rate,
        max_rate: commission.maxRate,
        max_change_rate: commission.maxChangeRate,
        special_property_here: commission.specialPropertyHere.map(el0 => ({
          rate: el0.rate,
          max_rate: el0.maxRate,
          max_change_rate: el0.maxChangeRate
        }))
      },
      str
    };
  },
  fromAmino: ({
    duration_camel_case,
    coins_camel_case,
    some_duration,
    some_height,
    my_coin,
    camel_case_name,
    vote_value,
    previous_votes,
    desc,
    commission,
    str
  }: AminoMstThing[\\"value\\"]): MstThing => {
    return {
      durationCamelCase: Long.fromString(duration_camel_case),
      coinsCamelCase: Array.from(coins_camel_case),
      someDuration: {
        seconds: Long.fromNumber(Math.floor(parseInt(some_duration) / 1_000_000_000)),
        nanos: parseInt(some_duration) % 1_000_000_000
      },
      someHeight: some_height ? {
        revisionHeight: Long.fromString(some_height.revision_height || \\"0\\", true),
        revisionNumber: Long.fromString(some_height.revision_number || \\"0\\", true)
      } : undefined,
      myCoin: my_coin,
      camelCaseName: camel_case_name,
      voteValue: voteOptionToJSON(vote_value),
      previousVotes: previous_votes.map(el => voteOptionToJSON(el)),
      desc: {
        moniker: desc.moniker,
        identity: desc.identity,
        website: desc.website,
        securityContact: desc.security_contact,
        details: desc.details,
        superNested: {
          rate: desc.super_nested.rate,
          maxRate: desc.super_nested.max_rate,
          maxChangeRate: desc.super_nested.max_change_rate,
          specialPropertyHere: desc.super_nested.special_property_here.map(el2 => ({
            rate: el2.rate,
            maxRate: el2.max_rate,
            maxChangeRate: el2.max_change_rate
          }))
        },
        manyComissions: desc.many_comissions.map(el1 => ({
          rate: el1.rate,
          maxRate: el1.max_rate,
          maxChangeRate: el1.max_change_rate,
          specialPropertyHere: el1.special_property_here.map(el2 => ({
            rate: el2.rate,
            maxRate: el2.max_rate,
            maxChangeRate: el2.max_change_rate
          }))
        }))
      },
      commission: {
        rate: commission.rate,
        maxRate: commission.max_rate,
        maxChangeRate: commission.max_change_rate,
        specialPropertyHere: commission.special_property_here.map(el1 => ({
          rate: el1.rate,
          maxRate: el1.max_rate,
          maxChangeRate: el1.max_change_rate
        }))
      },
      str
    };
  }
}"
`;

exports[`makeAminoTypeInterface 1`] = `
"export interface AminoMstThing extends AminoMsg {
  type: \\"cosmos-sdk/MstThing\\";
  value: {
    duration_camel_case: string;
    coins_camel_case: Coin[];
    some_duration: Duration[];
    some_height: AminoHeight;
    my_coin: Coin;
    camel_case_name: string;
    vote_value: number;
    previous_votes: number[];
    desc: {
      moniker: string;
      identity: string;
      website: string;
      security_contact: string;
      details: string;
      super_nested: {
        rate: string;
        max_rate: string;
        max_change_rate: string;
        special_property_here: {
          rate: string;
          max_rate: string;
          max_change_rate: string;
        }[];
      };
      many_comissions: {
        rate: string;
        max_rate: string;
        max_change_rate: string;
        special_property_here: {
          rate: string;
          max_rate: string;
          max_change_rate: string;
        }[];
      }[];
    };
    commission: {
      rate: string;
      max_rate: string;
      max_change_rate: string;
      special_property_here: {
        rate: string;
        max_rate: string;
        max_change_rate: string;
      }[];
    };
    str: string;
  };
}"
`;
